docs/features/nullable-reference-types.md

```diff
@@ -164,11 +164,59 @@ var z = (IEnumerable<object?>)x; // no warning
+
+When adding to the set of exact bounds of a type parameter, types that are equivalent (i.e. have an identity conversion) but not identical are merged using *invariant* rules.  When adding to the set of lower bounds of a type parameter, types that are equivalent but not identical are merged using *covariant* rules.  When adding to the set of upper bounds of a type parameter, types that are equivalent but not identical are merged using *contravariant* rules.  In the final fixing step, types that are equivalent but not identical are merged using *covariant* rules.
+
+Merging equivalent but not identical types is done as follows:
+
+#### Invariant merging rules
+
+Merging `dynamic` and `object` results in the type `dynamic`.
+Merging tuple types that differ in element names is specified elsewhere.
+Merging equivalent types that differ in nullability is performed as follows: merging the types Tn and Um (where n and m are differing nullability annotations) results in the type Vk where V is the result of merging T and U using the invariant rule, and k is as follows:
+ - if either n or m are non-nullable, non-nullable. In this case, if the other is nullable, a warning should be produced.
+ - if either n or m are nullable, nullable.
+ - otherwise oblivious.
+Merging constructed generic types is performed as follows: Merging the types K<A1, A2, ...> and K<B1, B2, ...> results in the type K<C1, C2, ...> where Ci is the result of merging Ai and Bi by the invariant rule.
+Merging the array types T[] and U[] results in the type V[] where V is the result of merging T and U by the invariant rule.
+
+#### Covariant merging rules
+
+Merging `dynamic` and `object` results in the type `dynamic`.
+Merging tuple types that differ in element names is specified elsewhere.
+Merging equivalent types that differ in nullability is performed as follows: merging the types Tn and Um (where n and m are differing nullability annotations) results in the type Vk where V is the result of merging T and U using the covariant rule, and k is as follows:
+ - if either n or m are nullable, nullable.
+ - if either n or m are oblivious, oblivious.
+ - otherwise non-nullable.
+Merging constructed generic types is performed as follows: Merging the types K<A1, A2, ...> and K<B1, B2, ...> results in the type K<C1, C2, ...> where Ci is the result of merging Ai and Bi by
+  - the invariant rule if the K's type parameter in the i position is invariant.
+  - the covariant rule if the K's type parameter in the i position is declared `out`.
+  - the contravariant rule if the K's type parameter in the i position is declared `in`.
+Merging the array types T[] and U[] results in the type V[] where V is the result of merging T and U by the covariant rule.
```

Comment:
Merging array element nullability currently uses the invariant rule. For instance, we report `"No best nullability"` for the following:
```c#
    static void F(object[] notNullableArray, object?[] nullableArray)
    {
        var x = new[] { notNullableArray, nullableArray };
    }
```